pkg/chunk/cache/memcached.go

```diff
@@ -113,37 +113,38 @@ func memcacheStatusCode(err error) string {
 }
 
 // Fetch gets keys from the cache. The keys that are found must be in the order of the keys requested.
-func (c *Memcached) Fetch(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string, err error) {
-	err = instr.TimeRequestHistogramStatus(ctx, "Memcache.Get", memcacheRequestDuration, memcacheStatusCode, func(ctx context.Context) error {
-		sp := opentracing.SpanFromContext(ctx)
-		sp.LogFields(otlog.Int("keys requested", len(keys)))
-		defer func() {
-			sp.LogFields(otlog.Int("keys found", len(found)), otlog.Int("keys missing", len(missed)))
-		}()
-
-		var err error
+func (c *Memcached) Fetch(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string) {
+	instr.TimeRequestHistogramStatus(ctx, "Memcache.Get", memcacheRequestDuration, memcacheStatusCode, func(ctx context.Context) error {
 		if c.cfg.BatchSize == 0 {
-			found, bufs, missed, err = c.fetch(ctx, keys)
-			return err
+			found, bufs, missed = c.fetch(ctx, keys)
+			return nil
 		}
 
-		found, bufs, missed, err = c.fetchKeysBatched(ctx, keys)
-		return err
+		found, bufs, missed = c.fetchKeysBatched(ctx, keys)
+		return nil
 	})
-
 	return
 }
 
-func (c *Memcached) fetch(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string, err error) {
+func (c *Memcached) fetch(ctx context.Context, keys []string) (found []string, bufs [][]byte, missed []string) {
 	var items map[string]*memcache.Item
-	err = UntracedCollectedRequest(ctx, "Memcache.GetMulti", instr.NewHistogramCollector(memcacheRequestDuration), memcacheStatusCode, func(_ context.Context) error {
+	err := instr.TimeRequestHistogramStatus(ctx, "Memcache.GetMulti", memcacheRequestDuration, memcacheStatusCode, func(_ context.Context) error {
+		sp := opentracing.SpanFromContext(ctx)
+		sp.LogFields(otlog.Int("keys requested", len(keys)))
+
 		var err error
 		items, err = c.memcache.GetMulti(keys)
+
+		sp.LogFields(otlog.Int("keys found", len(items)))
+		if err != nil {
+			sp.LogFields(otlog.Error(err))
+		}
 		return err
 	})
 
 	if err != nil {
 		missed = keys
+		level.Error(util.Logger).Log("msg", "Failed to get keys from memcached", "err", err)
```

Comment:
If we are trying to return partial results now shouldn't missed be the difference between keys and the items found?

Reply:
Yeah good point!